assayist/client/query.py

```diff
@@ -15,6 +15,31 @@ def set_connection(neo4j_url):
     neomodel.db.set_connection(neo4j_url)
 
 
+def get_container_by_component(component_name, component_version):
+    """Query for builds by component name and version.
+
+    Finds builds that produced container images which include or embed a component with the
+    specified name and version, and any version preceding that version.
+
+    :param str component_name: name of the component to query for
+    :param str component_version: version of the component to query for
+    :return: list of builds IDs
+    :rtype: list
+    """
+    query = """
+        MATCH (c:Component {{canonical_name: '{name}'}}) <-[:SOURCE_FOR]-
+              (:SourceLocation {{canonical_version: '{version}'}}) <-[:UPSTREAM|EMBEDS*0..]-
+              (target_sl:SourceLocation) <-[:BUILT_FROM]- (build:Build)
+        OPTIONAL MATCH (target_sl) <-[:SUPERSEDES*0..]- (sl:SourceLocation)
+        OPTIONAL MATCH (build) <-[:EMBEDS*]- (:Artifact) <-[:EMBEDS]- (target_sl)
+        RETURN build
```

Comment:
Unless I'm mistaken I think there's an addition case that's missing here: Builds that embed Artifacts that were built from the builds. Right now you're only trying to look for embedded Artifacts that embed the target_sl, not Artifacts that embed the produced Artifacts. Although maybe that's intentional and you're only looking for "first order" affected builds.